### PR TITLE
refactor: CLI commands for database management

### PR DESCRIPTION
**Pull request by polarathene**
# Description

In preparation for switching the file locking back to flock, I wanted to clean up the CLI commands involved, but got a bit carried away :sweat_smile: 

It should also help with keeping the encryption feature PR simple to resolve/test.

### Commands refactored:
- User (**All:** add / list / update / del + _dovecot-master variants_)
- Quota (**All:** set / del)
- Virtual Alias (**All:** add / list /del)
- Relay (**All:** add-relayhost / add-sasl / exclude-domain)

### Overall changes involve:
- **Fairly common structure:**
  - `_main` method at the top provides an overview of logical steps:
    - After all methods are declared beneath it (_and imported from the new `helpers/database/db.sh`_), the `_main` is called at the bottom of the file.
    - `delmailuser` additionally processes option support for `-y` prior to calling `_main`.
  - `__usage` is now consistent with each of these commands, along with the `help` command.
  - Most logic delegated to new helper scripts. Some duplicate content remains on the basis that it's low-risk to maintenance and avoids less hassle to jump between files to check a single line, usually this is arg validation.
  - Error handling should be more consistent, along with var names (_no more `USER`/`EMAIL`/`FULL_EMAIL` to refer to the same expected value_).
- **Three new management scripts** (in `helpers/database/manage/`) using a common structure for managing changes to their respective "Database" config file.
  - `postfix-accounts.sh` unified not only add and update commands, but also all the dovecot-master versions, a single password call for all 4 of them, with a 5th consumer of the password prompt from the relay command `addsaslpassword`.
  - These scripts delegate actual writes to `helpers/database/db.sh` which provides a common API to support the changes made.
     - This is more verbose/complex vs the current inline operations each command currently has, as it provides generic support instead of slightly different variations being maintained, along with handling some edge cases that existed and would lead to bugs (notably substring matches).
     - Centralizing changes here seems wiser than scattered about. I've tried to make it easy to grok, hopefully it's not worse than the current situation.
     - List operations were kept in their respective commands, `db.sh` is only really managing writes. I didn't see a nice way for removing the code duplication for list commands as the duplication was fairly minimal, especially for `listalias` and `listdovecotmasteruser` which were quite simple in their differences in the loop body.
     - `listmailuser` and `delmailuser` also retain methods exclusive to respective commands, I wasn't sure if there was any benefit to move those, but they were refactored.

Happy to drop the `helpers/database/**` files if it seems a bit over-engineered :+1: 

I figured with `flock` the locks can often be done in the `db.sh:_db_operation` method, only some methods may want to lock multiple files for larger operations such as `delmailuser`. This would be fine with `flock -x` in the same process, and not conflict with `flock -x` in `_db_operation` (_it'll re-use the existing lock for the FD it has instead of blocking_).

---

Apologies for the large commit history not likely being that helpful for review (_file diffs probably aren't great either_). I went through several iterations to reach this point.

If the PR changes are a bit large for review, let me know and I can split this out into smaller PRs.

<details>
<summary>Commit messages for reference</summary>

```
* refactor: `addmailuser` + `adddovecotmasteruser`

These two are very similar, updated together.

Restructures logic into functions, additional comments, local vars.

* refactor: `updatemailuser` + `updatedovecotmasteruser`

These two are very similar, updated together.

Restructures logic into functions, additional comments, local vars.

* refactor: `delmailuser` + `deldovecotmasteruser`

These two are very similar, updated together.

Restructures logic into functions, additional comments, local vars.

* refactor: `listmailuser`

Restructures logic into functions, additional comments, local vars.

Heavier refactor for this file. Dovecot Quota logic refactored, listing account aliases logic wrapped into it's own method.

`echo` output adjusted to not depend on conditional branches, but could be reverted.

* refactor: `listalias` + `listdovecotmasteruser`

These two are very similar, updated together.

Restructures logic into functions, additional comments, local vars.

* refactor: `addalias`

Restructures logic into functions, additional comments, local vars.

* refactor: `delalias`

Restructures logic into functions, additional comments, local vars.

* refactor: `setquota` + `delquota`

These two are very similar, updated together.

Restructures logic into functions, additional comments, local vars.

* refactor: Relay commands

Restructures logic into functions, additional comments, local vars.

* chore: Normalize to `MAIL_ACCOUNT` var

* chore: Corrections for `addalias` + `delalias`

`MAIL_ACCOUNT` usage and method names for these files were misunderstood, renaming to what they should be.

* chore: Update `sed` pattern

Should not use white-space as a delimiter in the pattern..

* chore: Use `_main` to position main logic at the top

* chore: `shellcheck` rule SC2904 no longer needed

* chore: Revise common method for relay commands

* chore: Sync common code `addmailuser` + `adddovecotmasteruser`

Majority of shared logic extracted into `add-account.sh` for maintenance.

* chore: Sync common code `updatemailuser` + `updatedovecotmasteruser`

Majority of shared logic extracted into `update-account.sh` for maintenance.

* chore: Align `addsaslpassword` with other password prompt methods

* chore: Share common helper `_if_missing_request_password`

* chore: Add helper `_hash_password`

* chore: Add helper `_account_already_exists`

* chore: Handle`_account_already_exists` errors in helper

* chore: Sync `MAIL_ACCOUNT` validation into helper method

* chore: Migrate dovecot quota methods to a common helper

* chore: Minor revisions

- Added some notes.
- Renamed some methods, `_password_hash` now takes args instead of using external vars.

* chore Minor revisions part 2

Bring back a common `_validate_paramters()` to command files for now.

Have `_account_already_exists` default to `postfix-accounts.cf` as that's the most commonly used DB for this method.

Renamed `_mail_account` helper methods to scope name to the context.

* chore: Use single helper file for accounts

* chore: Move common method for relay commands to helper method

* refactor: Migrate virtual alias methods into their own helper

- Logic restructured heavily in `listmailuser`, also fixing the quota regression I introduced (when moving the `ENABLE_QUOTA` guard into `_quota_show_for()`).
- `addalias` + `delalias` share common validation check in the new helper, their migrated methods is unaltered.
- The new helper adds some extra documentation.

* refactor: `listalias` + `listmailuser` + `listdovecotmasteruser`

Adds a common helper method to perform the DB check and list iterations while delegating a method for each to format the output per line/entry as they see fit.

This helper calls an internal method that would be defined earlier by each command. Similar to how `__usage` is called by several methods in `helper.sh` already.

* chore: `delmailuser` revisions

* refactor: `delmailuser` + `deldovecotmasteruser`

Methods updated to use explicit args instead of implicit ones. Bit easier to reason with currently despite the added verbosity/noise.

Added a `_key_exists_in_db` method to `helper.sh`, links common `grep -qi` lines that perform this lookup.

---

Refactor `delmailuser` DB entry removals to delegate to removal methods.  Rearranged the order, added inline docs explain why account removal occurs last, also moved maildir removal into this method.

`_alias_remove_for_recipient` is roughly the same logic as what `delmailuser` had for account removal from `postfix-virtual.cf`. Prefer this method of the version in `delmailuser`, but:
- Adopt the escaped inputs, so `.` only matches `.`, not _any_ character.
- Change white-space match of first expression to white-space token `\s` and expect one or more (`+`), not zero or more (`*`) to ensure there is a white-space separator.

Changed quota removal section to output a failure error like the other two removals do. Updated the quota helper method to accommodate by returning the status of sed (via sedfile) and not failing if no account exists to remove.

Account removal logic is shared/duplicated between `delmailuser` and `deldovecotmasteruser`. In `delmailuser` there is extra notes about account checking and info log that should perhaps be addressed.

Addressed the exit concern for `MAILDEL` handling in `_remove_maildir`, return seems to be the original intention.

* chore: Update `__usage` for each command

For the commands tackled by this PR, the `__usage` output has been normalized for consistency:
- Revisions to existing content such as in `DESCRIPTION` and `EXAMPLES`, renaming `SYNOPSIS` to `USAGE` and removal of `NAME` section.
- The command name at the top of the output is now lower-case.

`delmailuser` dropped `-h` option and `USAGE` syntax that accommodated the `help` option/sub-command.

`deldovecotmasteruser` has no options to parse, removed the OPTIND shift.

* fix: `setquota` and `delmailuser` (alias removal)

The original `setquota` prompt didn't actually prompt (`-p`), but used `-s` which silents the input, meant for passwords..

Added extra comments to clarify Dovecot quota value retrieved is in units of kibibytes, which is what we multiply by to get the total number of bytes for converting to IEC unit.

---

In `delmailuser`, the shared alias removal code I had it used did not have parity with the full entry removal pattern as a blank alias key was required (due to `^`), corrected by falling back to pattern matching non-whitespace (the alias key) for when the alias arg is empty.

* refactor: DB methods

`_key_exists_in_db` and `_db_add_or_replace_entry` now handle escaping of args internally. Relay commands were not escaping for regex pattern usage previously.

`_db_add_or_replace_entry` now takes an additional DELIMITER arg so that it can be used elsewhere (`setquota`).

`DELIMITER` is escaped for regex lookup if it's `|` (used for mail account delimiter) due to using extended regex. May consider reverting need for extended regex.

`_quota_exists_for_account` no longer needed.

* refactor: DB operations

- Manage DB delimiters in the main helper instead.
- Modified DB operations to several methods that forward args to single method using case statements with an action/operation to perform (reducing verbosity from duplicate operation code). This now additionally supports entry appending and removing.
- Updated helpers to use this new method instead (dropping redundant code), and existing ones to adapt to new args.
- Reverted extended regexp syntax, probably not relevant here? (ideally we forbid such special chars prior)

* refactor: DB operations alternative approach

- Changed back to separate KEY and VALUE args, removal operations would not have a delimiter to split from the ENTRY value otherwise.
- Fixed append operation for virtual-alias, missing the `,` prefix in it's VALUE.
- DATABASE moved to first arg to support remove operation better (avoids need to transpose args).
- Shortened db convenience method names, removing redundant suffix.
- This change removed need for the unpleasant ENTRY delimiter splitting for deriving KEY and VALUE. In exchange when composing ENTRY value we need to replace the DELIMITER for writing when it is `\s`.

* refactor: DB 'remove' + 'append' to support `postfix-virtual.cf`

The virtual alias DB handling has been moved into the helper for generic DB operations.

The logic was not ensuring exact matches for VALUE removal and appending (already exists check) leading to false-positives.

`_alias_list_for_account` still needs revision for proper exact matches.

* refactor: Unify account management

Create/Update/Delete methods can all leverage the same helper method, with convenience wrappers to simplify command API.

This seems better, commands have now inlined some methods, and may inline the parameter validation at a later point.

All operations are now using a lock from a single place, but this may better suit being moved into the main DB operation method, with the only concern being `delmailuser` may want to lock several files it touches? (regardless of loop concern).

* refactor: Unify quota management

Maybe not as useful for this helper due to only two methods and three consumers (setquota, delquota, delmailuser).

Methods used exclusively for `listmailuser` and `setquota` have been returned to those commands.

* refactor: Unify virtual alias management

Similar to previous commit for quota, the two methods with three consumers (addalias, delalias, delmailuser) adopt the same switch case approach.

Validation of args inlined. Args have the same order again, no longer supporting blank `MAIL_ALIAS`, instead expecting `_` to explicitly set the scope to all aliases.

`listmailuser` specific method returned to the command.

* chore: `_create_lock` is not reliable

In it's present form this only locks the command script itself, not a target file (Database) from other commands.

Kept it here as a reminder. It would be more appropriate now in `_db_operation`, however the current locking support would require more intricate lock releasing in that method, and would not be as ideal as locking the files prior to the loop starting..

* chore: Shift account validation into accounts helper

* chore: Return list iteration to commands

For similar reasons to allowing some duplicate code with relay commands, the list iterator was move into each list command to prefer readability.

Slightly improved the alias list conditional.

* chore: Minor revisions to relay commands

Validation shifted back.

The duplication is not as ideal for the common DOMAIN arg, but neither is having to reference it from a single file. 

MAIL_ACCOUNT and it's validation wasn't accurate message for the purpose of a relay account used as credential. Opted to inline as well.

* refactor: Minor revisions to `delmailuser`

Switch `MAILDEL` from boolean commands to 0 and 1 values, where 1 is true/enabled.

`MAILDEL` conditional message for `_remove_maildir` seemed redundant, especially in the context of a loop. May be of some use without loop support.

Folded mailbox directory check, anything after it assumes directory is available, so no conditional wrapping required.

* chore: Relocate DB helper scripts

* chore: Adjust imports

* chore: Minor revisions

- Imports did not work well for `db.sh`, mimic the `helpers/index.sh` approach instead.

- `_db_operation` had a typo with arg references.

- `_check_database_has_content` namespaced under `_db` prefix. Uses the `should` convention to convey expectation/validation intent. Similar was done for `_key_exists_in_db` which has different semantics (no failure built-in), thus avoids `should`. Both moved down to bottom of file as validation focused.

- `_db_get_delimiter_for` pushed further down in the file, added clarification it is an internal method now.

- `_password_hash` is only used in one location now, not 4, inlined it.

- Migrated `_password_request_if_missing` to `postfix-accounts.sh` as that's the main consumer, with only `addsaslpassword` also using it.

- Additional context comments on consumers of some methods.

* fix: DB operation corrections

Better distinguish between DELIMITER for KEY and VALUE (may differ for lists or writing value) with `K_` and `V_` prefixes.

Sync'd duplicate block in `_db_operation` and `_db_has_entry_with_key`.

---

`replace` for ENTRY was breaking for `updatemailuser` due to sed replacment var containing `/` (password base64 encoded), where previously the method had sed use ` ` (space) as a sed  substitution command delimiter.

`append` does not use the new `__escape_sed_replacement` method for `VALUE`, although perhaps it should for consistency?

---

Removing value from a list failed an alias test. There was no condition to match the left-most value in a list of multiple items.

Fixed by using an OR condition on the third expression, which it will collapse to the value captured (`\1`). This can be `\s` and the matched white-space value will be the captured value.

---

`__db_list_already_contains_value` doesn't seem to need `-n` + `p`? I cannot think of a scenario where the logic breaks now.

If it does, revert logic back, and add a third sed expression before grep that should output single values when no delimiter is present `-e "/^\S\+$/p"`. Previously single value items were not matched..

Revised the inline docs for the method.

* fix: Create config dir for databases if it doesn't exist

Only during an "add" fallback operation.

* tests: Make them pass

Current behaviour introduced was to error that the user does not exist, even if the database does not exist.

For now satisfying the test that expects a soft failure without an error.

* lint: Appease the lint gods with an offering of guidance

`source-path` added to `index.sh` is sufficient if used directly above the `source` line, otherwise it can be placed early on in the sourced file, but must be before the dependent `source` lines appear (function scope is ignored).

Likewise, I forgot that the `PATH_TO_SCRIPTS` var is only for a common ancestor of the real path and shellchecks source path, it's stripped/ignored by shellcheck in favor of trying any known source-path. So `/manage` must be added to each of these `source` lines instead.

The prior `source-path` applied within the function, as noted only applied to the direct line beneath it, not the entire scope.

Added comments for future maintainers to be aware of. Decided to keep it all within this `db.,sh` file, rather than `helpers/index.sh` , although perhaps that's a better place to document/demonstrate such. Otherwise `test/linting/lint.sh` covers this advice for the most part.
```

</details>

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by polarathene — COMMENTED** [review]
Additional commentary if it's helpful.

Just to clarify. Multi-line values in .cf` files is not supported, nor is resolving recursive aliases. This just brings better consistency and reliability to the current support.

File: target/scripts/helpers/database/db.sh [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  local V_DELIMITER="${K_DELIMITER}"
+  [[ ${V_DELIMITER} == '\s' ]] && V_DELIMITER=' '
+  local ENTRY="${KEY}${V_DELIMITER}${VALUE}"
+
+  # Support for 'append' + 'remove' operations on value lists:
+  # NOTE: Presently only required for `postfix-virtual.cf`.
+  local _VALUE_
+  _VALUE_=$(_escape "${VALUE}")
+  # `postfix-virtual.cf` is using `,` for delimiting a list of recipients:
+  [[ ${DATABASE} == "${DATABASE_VIRTUAL}" ]] && V_DELIMITER=','
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/helpers/database/db.sh:63]: I actually like it. If in doubt, I'd go with more verbose than less verbose, for "future generations" :D

File: target/scripts/helpers/database/db.sh [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+function __db_list_already_contains_value
+{
+  # Avoids accidentally matching a substring (case-insensitive acceptable):
+  # 1. Extract the current value of the entry (`\1`),
+  # 2. If a value list, split into separate lines (`\n`+`g`) at V_DELIMITER,
+  # 3. Check each line for an exact match of the target VALUE
+  sed -e "s/^${KEY_LOOKUP}\(.*\)/\1/" \
+      -e "s/${V_DELIMITER}/\n/g"      \
+      "${DATABASE}" | grep -qi "^${_VALUE_}$"
+}
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/helpers/database/db.sh:140]: So should we change `listmailuser` to use this as well?

**Review by georglauterbach — APPROVED** [review]
So, I took some time to look at everything, but this is a rather large PR, so there is a fair chance of missing something. But as far as I can tell, LGTM 👍🏼

File: target/scripts/helpers/database/db.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (polarathene) [reply target/scripts/helpers/database/db.sh:140]: > So should we change `listmailuser` to use this as well?

The concern is raised in the [`listmailuser` review comment](https://github.com/docker-mailserver/docker-mailserver/pull/2654#discussion_r902415967). A check that is more robust won't be as helpful if the actual logic that is performed is problematic itself :sweat_smile:

**Comment by polarathene** [comment]
If someone else is willing to review, just let me know if you'd prefer to review this as multiple separate PRs instead and I'll split it up for you, no worries :+1:

**Comment by casperklein** [comment]
I'll review this hopefully in the next week. I prefer reviewing in one take. So big PRs needs a time frame that big enough. Smaller PRs I can review more occasionally.

**Comment by casperklein** [comment]
@polarathene I didn't forget about this, but it's really hard for me at the moment to find some spare time. I will review asap.

**Comment by polarathene** [comment]
> @polarathene I didn't forget about this, but it's really hard for me at the moment to find some spare time.

No worries :+1: 

If you'd prefer I can always try split it up into smaller PRs. I've still got about a month left that I can continue to contribute for before I'm swamped again. PRs after this one should be smaller.

**Comment by georglauterbach** [comment]
Feel free to merge this when approved. However, you'd need to switch all issues / PR from milestone `v11.2.0` to `v12.0.0` for completeness :)

**Comment by polarathene** [comment]
> However, you'd need to switch all issues / PR from milestone `v11.2.0` to `v12.0.0` for completeness :)

Any reason for a v12 release?

I wanted to tackle the encryption PR which seemed fine for 11.2, then the other WIP PRs would likely be breaking for a v12.

If you don't think an 11.2 release is worthwhile, that's fine. I was hoping to squeeze in some other internal improvements, but now I might not have enough time :sweat_smile:

**Comment by georglauterbach** [comment]
> > However, you'd need to switch all issues / PR from milestone `v11.2.0` to `v12.0.0` for completeness :)
> 
> Any reason for a v12 release?

No, but you marked this for v12, so I figured we'd need the v12 release. If not, we should mark this for v11.2.0.

**Comment by georglauterbach** [comment]
Maybe splitting this into smaller PRs is a good option now 😅

**Review by casperklein — APPROVED** [review]
> So, I took some time to look at everything, but this is a rather large PR, so there is a fair chance of missing something. But as far as I can tell, LGTM 👍🏼

I second this. Also the tests all pass, so I finally approve this 🎉

**Comment by polarathene** [comment]
Thanks to both of you for taking the time to review, again apologies for the size, I realize how daunting that'd be to review :sweat_smile:

**Pull request merged**